### PR TITLE
Networking fixes

### PR DESCRIPTION
**Pull request by ReubenBond**
* Fix wire protocol backwards compatibility with older versions
* Implement most feedback from #5436
* Add protocol version option for upgradeability of network protocol
* General cleanup in networking code

### CODE REVIEW

File: src/Orleans.Runtime/Networking/SiloConnection.cs [hunk 1]
```diff
@@ -158,20 +165,46 @@ protected override async Task RunInternal()
...
             finally
             {
-                if (!(this.remoteSiloAddress is null)) this.connectionManager.OnConnectionTerminated(this.remoteSiloAddress, this);
+                if (!(this.RemoteSiloAddress is null)) this.connectionManager.OnConnectionTerminated(this.RemoteSiloAddress, this);
             }
 
-            Task WritePreamble() => ConnectionPreamble.Write(this.Context, Constants.SiloDirectConnectionId, this.myAddress);
+            async Task WritePreamble()
+            {
+                await ConnectionPreamble.Write(
```
Comment (JanEggers) [inline src/Orleans.Runtime/Networking/SiloConnection.cs]: this does not need to be async just return the task

File: src/Orleans.Runtime/Networking/SiloConnectionFactory.cs [hunk 2]
```diff
@@ -47,7 +49,24 @@ protected override Connection CreateConnection(ConnectionContext context)
...
+
+        private void EnsureInitialized()
+        {
+            if (!isInitialized)
+            {
+                lock (this.initializationLock)
+                {
+                    if (!isInitialized)
+                    {
+                        this.messageCenter = this.serviceProvider.GetRequiredService<MessageCenter>();
```
Comment (JanEggers) [inline src/Orleans.Runtime/Networking/SiloConnectionFactory.cs]: why cant you use  ctor injection for these services?

Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/SiloConnectionFactory.cs]: Circular dependency

File: src/Orleans.Runtime/Networking/SiloConnection.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.Runtime/Networking/SiloConnection.cs]: It's a ValueTask

**Comment by ReubenBond** [comment]
I'm splitting this up into smaller PRs

**Pull request closed without merging**